### ISSUE TITLE
Proposal: Support running a quorum of validator clients with leader election

### ISSUE THREAD

**Issue by ThomasDalla** [Consensys-Incorporated/teku#6240]
### Rationale
In most clients, including Teku, the validator client is a single point of failure. Even if you are using distributed remote signers and multiple EL/CL pairs of nodes, the validator client remains in most infrastructures a single point of failure.
While there are solutions coming up to achieve high resiliency (DVT), this is early stage and requires a fundamental infrastructure change.
Being able to run Teku validator clients in a reliable active/standby fashion would remove the single point of failure, making it a lot more resilient. 

### Implementation
One low-effort high-benefit way to implement an active-standby failover for Teku validator client is to use the [Apache Curator Leader Election recipe](https://curator.apache.org/curator-recipes/leader-latch.html), that relies on a preferably replicated [Apache Zookeeper quorum](https://zookeeper.apache.org/doc/r3.1.2/zookeeperStarted.html#sc_RunningReplicatedZooKeeper), whose setup is beyond the scope of Teku and assumed already configured and running.
From Teku's perspective, we only need to specify the quorum connection string (example: `127.0.0.1:2181`), the path for the group (optional, defaulted to `/validator-client`) and the participant ID (example: `1`, `2`, `3` for each validator client).
This approach would also make the quorum agnostic to a particular client, meaning we could have a quorum with a mix of different validator clients, increasing further the diversity and therefore resiliency (if we ever implement that feature on other clients).

### Risks
With the right, conservative configuration of the `sessionTimeout` (default: 60 sec) and `connectionTimeout` (default: 15 sec), there should be very little to no risk of having two validator clients being up and asserting at the same time, but in order to protect from an unforeseen edge case issue, using a **distributed remote signer** is recommended (as it would ensure 2 validator clients cannot assert/propose the same message).

### Draft
To illustrate how simple the changes would be to support quite a powerful feature, I have committed a draft: https://github.com/ThomasDalla/teku/commit/42705758cdc24eee9ec535c7192cfc6f2aceda5a

I have omitted the configuration part, focusing on the actual implementation that would require 2 small changes:

#### Teku node startup
Wrap the Node within a QuorumNode that runs only when we are the leader.
```java
    if (config.quorumConfig() != null
        && !Strings.isNullOrEmpty(config.quorumConfig().connectionString())) {
      QuorumNode quorumNode = new QuorumNode(node, config.quorumConfig());
      quorumNode.start();
      return quorumNode;
    } else {
      node.start();
      return node;
    }
```

#### QuorumNode implementation
`QuorumNode` implements `Node` (`start()`/`stop()`) and `LeaderLatchListener` (`isLeader()`, `notLeader()`).
```java
public class QuorumNode implements Node, LeaderLatchListener {
  private final Node node; // reference to the actual node we run when we are the leader
  private final CuratorFramework client; // the Zookeeper client
  private final LeaderLatch latch; // the latch that notifies us when we're leader or not

  public QuorumNode(Node node, QuorumConfig config) {
    this.node = node;
    this.client =
        CuratorFrameworkFactory.newClient(
            config.connectionString(),
            config.sessionTimeoutMs(),
            config.connectionTimeoutMs(),
            new ExponentialBackoffRetry(config.baseSleepTimeMs(), config.maxRetries()));
    this.latch = new LeaderLatch(client, config.getPath(), config.getParticipantId());
  }

  /** LeaderLatchListener.isLeader() | We acquired leadership, so let's start the node */
  @Override
  public void isLeader() {
    this.node.start();
  }

  /** LeaderLatchListener.notLeader() | We lost leadership, stop the node! */
  @Override
  public void notLeader() {
    this.node.stop();
  }

  /** Node.start() | Kick-off the process */
  @Override
  public void start() {
    this.client.start();
    try {
      this.latch.start();
    } catch (Exception e) {
      throw new UnsupportedOperationException(e);
    }
    Runtime.getRuntime().addShutdownHook(new Thread(this::stop));
  }

  /** Node.stop() | Exit the node (release the latch and stop the quorum client) */
  @Override
  public void stop() {
    this.node.stop();
    CloseableUtils.closeQuietly(this.latch);
    CloseableUtils.closeQuietly(this.client);
  }

  @Override
  public ServiceControllerFacade getServiceController() {
    return this.node.getServiceController();
  }
}
```

### Next steps
* Finish implementation (wire the configuration into command-line / env var / yaml file configs)
* Add unit tests
* Add metrics of the quorum's health
* Document the feature on Teku's doc

Feedback welcome!
Planning to test on Goerli as soon as I complete the implementation.

**Comment by benjaminion** [comment]
Hi @ThomasDalla - how are things going with this? It is an interesting approach. We're mainly focusing our attention on supporting the more standardised DVT platforms, but are open to features such as this.

**Comment by ThomasDalla** [comment]
Hi @benjaminion - thanks for your comment.
I am also following closely the standardised DVT platform (participating in their testnets), which is an exciting area.
This proposal is using a more common technology in the industry (Zookeeper) that many non-crypto developers are already familiar with, which might attract some companies and solo stakers.

I had a working prototype on Goerli that was running well, but never took the time to polish the code and add unit tests so nowhere close to a pull request level.
It highlighted 2 main points:
1. It only makes sense and should only be allowed if running a standalone validator client
2. When a new leader is elected, it starts performing duties immediately, so it should either force doppelganger detection (to ensure 2 epochs wait) or force the use of an external signer (like `web3signer`, which is the option I went for as doppelganger was not implemented yet in Teku when I experimented with that)

I have since taken the idea a step further and built a generic orchestration layer for docker containers that allows me to run a quorum of services, ensuring only the elected leader actually runs (other ones on standby).

The reason for going that way was to achieve client diversity without having to change the code of each client.
This is closer to what the standardised DVT platforms achieve.
If I want a Zookeeper quorum with a mix of different validator clients, I need to implement the feature as a first class citizen on each client (Teku, Prysm, Nimbus, Lodestar, etc...). As I barely have time and expertise (Java) to do so for one of them (Teku, ma favourite!), I would never be able to implement that natively in all of them (alone).

However, with an independent process that orchestrates the docker container of the validator client, it is agnostic to the code of the client itself.
The code is rather lightweight (a handful of classes) and consists in:
- a config that specifies the container (for the Docker API to orchestrate it) and the zookeeper endpoint
- joining the quorum and, if we are the leader, start the container, otherwise stop it
- in the background, periodically check that if we are the leader, the container is running and if we're not, the container is stopped (but exists)
- a Pushover config to alert me when an incident occurs (like auto-switching to a new host)

The periodic check is for additional protection as the zookeeper connection is "live" and we "receive" `obtainedLeadership()` / `lostLeadership()` events.
It's also to detect if the container is started by mistake (manually for example) while we are not the leader.
It also supports a "transition" mode where I can force switching to another leader when I want to do maintenance on the box that has leadership.

I have been running like that on Mainnet for months now and it has prevented me from going offline during power or network outages (all my boxes are solo boxes in various locations at family homes).

This diagram illustrates the setup (and my poor LibreOffice Draw skills):
![image](https://user-images.githubusercontent.com/609945/227633907-099de9e0-44a2-4a6e-ae84-64a89fa0cc1f.png)

And this is what the logs look like (periodic check) with hosts 1 and 2 (3 was down for maintenance):
![image](https://user-images.githubusercontent.com/609945/227633430-ca02edfb-db49-49f3-ada9-cde2e992492b.png)

Validating on Mainnet for months with only one downtime of about a minute for an upgrade of  `web3signer` that required a database update (and therefore to stop all of them).

On a closing note, I still think that supporting the quorum natively in Teku (and other clients) could have benefits (and why not, start a trend in other validator clients!) by removing the custom "docker-leader" layer.

If I find the time/motivation, I'll polish the code and submit something.
